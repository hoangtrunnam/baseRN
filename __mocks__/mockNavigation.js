/* eslint-disable semi */
const mockNavigation = {
  getUniqueId: jest.fn(() => 'uniqueId'),
  getVersion: jest.fn(() => '1.0'),
  hasNotch: jest.fn(() => false),
  navigate: jest.fn(),
  goBack: jest.fn(),
}

export default mockNavigation
