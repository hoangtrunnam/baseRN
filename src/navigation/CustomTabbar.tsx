/* eslint-disable react-native/no-inline-styles */
import type {BottomTabBarProps} from '@react-navigation/bottom-tabs'
import React from 'react'
import {SafeAreaView, StyleSheet, View} from 'react-native'
import Tab from './Tab'
import {DIMENSION} from 'src/commons/dimension'
import {defaultColors} from 'src/configs/colors'

export const BOTTOMTAB_HEIGHT = 68

const CustomTabbar = ({state, descriptors, navigation}: BottomTabBarProps) => {
  return (
    <SafeAreaView>
      <View style={[styles.container]}>
        <View style={{position: 'absolute', bottom: -8, left: 0, right: 0}}>
          <View style={styles.shadow}>
            <View
              style={{
                width: '100%',
                height: DIMENSION.isIos ? BOTTOMTAB_HEIGHT : 53,
                backgroundColor: defaultColors.h_ffffff,
              }}
            />
          </View>
        </View>
        <View style={styles.tabContainer}>
          {state.routes.map((route: any, index: any) => (
            <Tab
              key={index}
              {...{
                route,
                isFocused: state.index === index,
                options: descriptors[route.key].options,
                navigation,
              }}
            />
          ))}
        </View>
      </View>
    </SafeAreaView>
  )
}

export default CustomTabbar

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    // backgroundColor: 'gray',
    // flex: 1,
  },
  tabContainer: {
    marginTop: 'auto',
    width: '100%',
    height: 45,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    marginBottom: DIMENSION.hasNotch ? 15 : 0,
    backgroundColor: '#ffffff',
  },
  shadow: {
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: -2,
    },
    shadowOpacity: 0.15,
    shadowRadius: 3.84,
    elevation: 3,
  },
})
