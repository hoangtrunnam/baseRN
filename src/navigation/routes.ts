import type {RoutesType} from './types';

type Entries<T> = {
  [K in keyof T]: K;
};

export const routes: Entries<RoutesType> = {
  Notifications: 'Notifications',
  BottomTab: 'BottomTab',
  // Screen moi tu day
  // Auth
  SignUp: 'SignUp',
  Gift: 'Gift',
  Home: 'Home',
  Login: 'Login',
  UserProfile: 'UserProfile',
};
