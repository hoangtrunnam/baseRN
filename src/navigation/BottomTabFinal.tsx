import {createBottomTabNavigator} from '@react-navigation/bottom-tabs';
import React from 'react';
import CustomTabbar from './CustomTabbar';
import {routes} from './routes';
import type {BottomTabNavigationParamList} from './types';
import Home from '../features/TestApp/Home';
import Gift from '../features/TestApp/Gift';
import Notifications from '../features/TestApp/Notifications';
import UserProfile from '../features/TestApp/UserProfile';
import {useDisableBackRoute} from '../commons/hooks/useDisableBackRoute';

const Tab = createBottomTabNavigator<BottomTabNavigationParamList>();

const BottomTabFinal = () => {
  useDisableBackRoute();
  return (
    <Tab.Navigator
      detachInactiveScreens
      // eslint-disable-next-line react/no-unstable-nested-components
      tabBar={props => <CustomTabbar {...props} />}
      screenOptions={{
        headerShown: false,
        tabBarActiveTintColor: '#000234',
        tabBarInactiveTintColor: '#993243',
      }}
      initialRouteName={routes.Home}>
      <Tab.Screen name={routes.Home} component={Home} />
      <Tab.Screen name={routes.Gift} component={Gift} />
      <Tab.Screen name={routes.Notifications} component={Notifications} />
      <Tab.Screen name={routes.UserProfile} component={UserProfile} />
    </Tab.Navigator>
  );
};

export default BottomTabFinal;
