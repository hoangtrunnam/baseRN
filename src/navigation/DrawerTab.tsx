/* eslint-disable react/no-unstable-nested-components */
import {createDrawerNavigator} from '@react-navigation/drawer'
import React from 'react'
import {useDisableBackRoute} from 'src/commons/hooks/useDisableBackRoute'
import {routes} from './routes'
import CustomDrawer from './Component/CustomDrawer'
import Dashboard from 'src/features/Dashboard'

const Drawer = createDrawerNavigator()

const DrawerTab = () => {
  useDisableBackRoute()
  return (
    <Drawer.Navigator
      initialRouteName={routes.Dashboard}
      drawerContent={props => <CustomDrawer {...props} />}
      screenOptions={{
        headerShown: false,
        // swipeEdgeWidth: 0,
        drawerStyle: {
          width: '75%',
        },
        // drawerType: 'front',
        // swipeEdgeWidth: 0,
        // swipeEnabled: false,
      }}>
      <Drawer.Screen name={routes.Dashboard} component={Dashboard} />
    </Drawer.Navigator>
  )
}

export default DrawerTab
