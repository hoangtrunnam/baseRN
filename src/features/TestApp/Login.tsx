/* eslint-disable react-native/no-inline-styles */
import {View, Text, TouchableOpacity} from 'react-native'
import React from 'react'
import type {MainStackScreenNavigationProps} from 'src/navigation/types'
import {routes} from 'src/navigation/routes'

interface Login extends MainStackScreenNavigationProps<'Login'> {}

const Login = ({navigation}: Login) => {
  return (
    <View style={{flex: 1, justifyContent: 'center', alignItems: 'center'}}>
      <Text>Login</Text>
      <TouchableOpacity
        onPress={() => {
          navigation.navigate(routes.BottomTab, {screen: routes.Home})
        }}>
        <Text>go to main screen</Text>
      </TouchableOpacity>
    </View>
  )
}

export default Login
