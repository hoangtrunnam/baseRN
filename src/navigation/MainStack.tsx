import {createStackNavigator, TransitionPresets} from '@react-navigation/stack'
import React from 'react'
import {Platform} from 'react-native'
import {routes} from './routes'
import type {MainNavigationParamList} from './types'
import Gift from 'src/features/TestApp/Gift'
import Home from 'src/features/TestApp/Home'
import Notifications from 'src/features/TestApp/Notifications'
import UserProfile from 'src/features/TestApp/UserProfile'
import DrawerTab from './DrawerTab'
import DetailCalender from 'src/features/TestApp/DetailCalender'
import Login from 'src/features/TestApp/Login'
import BottomTabFinal from './BottomTabFinal'
const Stack = createStackNavigator<MainNavigationParamList>()

const MainStack: React.FC<any> = () => {
  const transitionPresets = Platform.OS === 'android' ? TransitionPresets.FadeFromBottomAndroid : {}
  return (
    <Stack.Navigator
      screenOptions={{
        headerShown: false,
        headerStyle: {
          shadowColor: 'transparent',
        },
        ...transitionPresets,
      }}
      initialRouteName={routes.Login}>
      <Stack.Screen
        name={routes.DrawerTab}
        component={DrawerTab}
        options={{headerShown: false, gestureEnabled: false}}
      />
      <Stack.Screen
        name={routes.BottomTab}
        component={BottomTabFinal}
        options={{headerShown: false, gestureEnabled: false}}
      />
      <Stack.Screen name={routes.Notifications} component={Notifications} />
      <Stack.Screen name={routes.Gift} component={Gift} />
      <Stack.Screen name={routes.Home} component={Home} />
      <Stack.Screen name={routes.UserProfile} component={UserProfile} />
      <Stack.Screen name={routes.DetailCalender} component={DetailCalender} />
      <Stack.Screen name={routes.Login} component={Login} />
    </Stack.Navigator>
  )
}

export default MainStack
