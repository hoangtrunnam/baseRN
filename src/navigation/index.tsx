import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
// import { useFlipper } from '@react-navigation/devtools'
import {refNavigation} from './navigationHelper';
import MainStack from './MainStack';

const Navigator = () => {
  // @ts-ignore
  // useFlipper(refNavigation)

  return (
    <NavigationContainer ref={refNavigation}>
      <MainStack />
    </NavigationContainer>
  );
};

export default Navigator;
