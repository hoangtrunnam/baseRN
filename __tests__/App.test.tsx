/**
 * @format
 */

import 'react-native'
import React from 'react'
import {Provider} from 'react-redux'
import rootReducer from '../src/redux/reducer'
import {useNavigation} from '@react-navigation/native'
// Note: import explicitly to use the types shiped with jest.
import {it} from '@jest/globals'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'
// import Gift from 'src/features/TestApp/Gift'
import {routes} from 'src/navigation/routes'
import mockNavigation from '../__mocks__/mockNavigation'
import {configureStore} from '@reduxjs/toolkit'
import Login from 'src/features/TestApp/Login'

jest.mock('@react-native-async-storage/async-storage', () => ({
  setItem: jest.fn((key, value) => Promise.resolve()),
  getItem: jest.fn(key => Promise.resolve(null)),
  removeItem: jest.fn(key => Promise.resolve()),
  clear: jest.fn(() => Promise.resolve()),
  getAllKeys: jest.fn(() => Promise.resolve([])),
}))

jest.mock('react-native-keychain', () => {
  // Cung cấp một mock implementation tại đây nếu cần
})
jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock')
  Reanimated.default.call = () => {}
  return Reanimated
})

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: jest.fn(),
}))

useNavigation.mockReturnValue(mockNavigation)

it('renders correctly', () => {
  const store = configureStore({
    reducer: rootReducer,
  })
  const component = renderer.create(
    <Provider store={store}>
      <Login navigation={mockNavigation as any} route={routes as any} />
    </Provider>
  )

  expect(component).toMatchSnapshot()
})
