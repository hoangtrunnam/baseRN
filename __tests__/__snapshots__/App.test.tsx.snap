// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
[
  <RCTSafeAreaView
    style={
      {
        "backgroundColor": "#ffffff",
      }
    }
  >
    <View
      leftIconType="none"
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "space-between",
          "marginTop": -10,
          "paddingHorizontal": 16,
          "width": "100%",
        }
      }
      title="Login screen"
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      />
      <View
        style={
          [
            {
              "height": "100%",
              "justifyContent": "center",
              "left": 375,
              "position": "absolute",
              "top": 0,
            },
            {
              "transform": [
                {
                  "translateX": -310,
                },
              ],
              "width": 620,
            },
          ]
        }
      >
        <Text
          style={
            {
              "alignItems": "stretch",
              "color": "#535353",
              "fontSize": 20,
              "fontWeight": "500",
              "textAlign": "center",
            }
          }
        >
          Login screen
        </Text>
      </View>
      <View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "transparent",
            }
          }
        >
          <Text
            style={
              [
                {
                  "alignItems": "center",
                  "color": undefined,
                  "fontSize": 16,
                  "fontWeight": "600",
                  "justifyContent": "center",
                  "textAlign": "center",
                  "textDecorationLine": "underline",
                },
                undefined,
                undefined,
              ]
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "#CBCBCB",
          "height": 1,
          "width": "100%",
        }
      }
    />
  </RCTSafeAreaView>,
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="24"
      bbWidth="24"
      fill="none"
      focusable={false}
      height={24}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      vbHeight={18}
      vbWidth={19}
      width={24}
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          clipRule={0}
          d="M16.6611 3.14668H14.3424L12.8636 0.592979C12.6018 0.141 11.9526 -0.211776 11.3266 0.148534L6.11808 3.14668H1.25143C0.397825 3.14668 0 3.9263 0 4.50191V16.6448C0 17.2204 0.397825 18 1.25143 18H16.6611C17.5147 18 17.9126 17.2204 17.9126 16.6448V13.4464H17.9937C18.4984 13.4464 19 13.0549 19 12.4556V8.47429C19 7.87497 18.4984 7.48341 17.9937 7.48341H17.9126V4.50191C17.9126 3.9263 17.5147 3.14668 16.6611 3.14668ZM11.8186 13.4464H16.2814V16.3737H1.63116V4.77295H16.2814V7.48341H11.8186C11.3139 7.48341 10.8123 7.87497 10.8123 8.47429V12.4556C10.8123 13.0549 11.3139 13.4464 11.8186 13.4464ZM11.6892 1.81765L12.4589 3.14668H9.3804L11.6892 1.81765ZM12.4435 9.10969V11.8201H17.3688V9.10969H12.4435Z"
          fill={
            {
              "payload": 4279114821,
              "type": 0,
            }
          }
          fillRule={0}
          propList={
            [
              "fill",
              "fillRule",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        {
          "height": 200,
          "width": 200,
        }
      }
    >
      <View
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "height": 200,
              "width": 200,
            },
          ]
        }
      >
        <FastImageView
          defaultSource={null}
          onFastImageError={[Function]}
          onFastImageLoad={[Function]}
          resizeMode="cover"
          source={null}
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0,0,0,0.4)",
            "borderBottomLeftRadius": undefined,
            "borderBottomRightRadius": undefined,
            "borderRadius": undefined,
            "borderTopLeftRadius": undefined,
            "borderTopRightRadius": undefined,
            "bottom": 0,
            "justifyContent": "center",
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <ActivityIndicator
          color="#62267B"
          size="small"
        />
      </View>
    </View>
    <Text>
      Login screen
    </Text>
    <View
      style={
        {
          "backgroundColor": "red",
          "borderRadius": 16,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={
          {
            "bottom": 10,
            "left": 10,
            "right": 10,
            "top": 10,
          }
        }
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onLayout={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text>
           Login
        </Text>
        <View
          pointerEvents="box-only"
          style={
            {
              "bottom": 0,
              "left": 0,
              "overflow": "hidden",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
    </View>
  </View>,
]
`;
